tutorial/lesson_22_jit_performance.cpp

```diff
@@ -0,0 +1,263 @@
+    {
+        size_t count = 0;
+        double t = benchmark(samples, iterations, [&]() {
+
+            // First, create an output buffer to hold the results.
+            Buffer<uint16_t> result(1024, 1024);
+            
+            // Now, construct our pipeline from scratch.
+            Pipeline pipeline = make_pipeline();
+
+            // And then call realize to execute the pipeline.
+            pipeline.realize(result);
+            ++count;
+        });
+
+        // On a MacBook Pro M1, we should get around ~1800 times/sec.
+        std::cout << "Compile & Execute Pipeline (from scratch): " << int(count / t) << " times/sec\n";
+    }
+
+    // This time, let's create the pipeline outside the timing loop and re-use it for each execution ...
+    {
+        // Create our pipeline, and re-use it in the loop below
+        Pipeline pipeline = make_pipeline();
+
+        size_t count = 0;
+        double t = benchmark(samples, iterations, [&]() {
+
+            // Create our output buffer
+            Buffer<uint16_t> result(1024, 1024);
+            
+            // Now, call realize
+            pipeline.realize(result);
+            ++count;
+        });
+
+        // On a MacBook Pro M1, we should get around ~175000 times/sec (almost 95-100x times faster!).
+        std::cout << "Compile & Execute Pipeline (re-use pipeline): " << int(count / t) << " times/sec\n";
+    }
+
+    // Let's do the same thing as before, but explicitly JIT compile before we realize ...
+    {
+        Pipeline pipeline = make_pipeline();
+
+        // Let's JIT compile for our target before we realize, and see what happens ...
+        const Target target = get_jit_target_from_environment();
+        pipeline.compile_jit(target);
+
+        size_t count = 0;
+        double t = benchmark(samples, iterations, [&]() {
+            Buffer<uint16_t> result(1024, 1024);
+            pipeline.realize(result);
+            ++count;
+        });
+ 
+        // On a MacBook Pro M1, this should be about the same as the previous run (about ~175000 times/sec)
+        //
+        // This may seem somewhat surprisingly, since compiling before realizing doesn't seem to make 
+        // much of a difference to the previous case.  However, the first call to realize() will implicitly
+        // JIT-compile and cache the generated code associated with the Pipeline object, which is basically 
+        // what we've done here. Each subsequent call to realize uses the cached version of the native code, 
+        // so there's no additional overhead, and the cost is amortized as we re-use the pipeline.
+        std::cout << "Execute Pipeline (compile before realize): " << int(count / t) << " times/sec\n";
+
+
+        // Another subtlety is the creation of the result buffer ... the declaration implicitly
+        // allocates memory. Instead, we could use the realize({1024, 1024}) call which use the
+        // buffer managed by the pipeline object for the outputs ...
```

Comment:
Another tiny style nit here and elsewhere: I think it's more correct to omit the space before an ellipsis, i.e., `for the outputs...` rather than `for the outputs ...`